### ISSUE TITLE
width="100%" leads to width="0"

### ISSUE THREAD

**Issue by titoBouzout** [ryansolid/mobx-jsx#33]
If you inspect the source on this demo you will see the width is 0 
https://codesandbox.io/s/width-attribute-is-0-rrs9g?file=/index.js

I dont know typescript but I figured this may have something to do with it, checking this file on
https://github.com/ryansolid/mobx-jsx/blob/master/src/jsx.ts#L1805 and 
https://github.com/ryansolid/mobx-jsx/blob/master/src/jsx.ts#L2069 

it seems to be declared as `width?: number | string;` or `width?: CSSWideKeyword | any;`, so I'm at a loss of what's going on.

**Comment by titoBouzout** [comment]
I forgot to mention, that I wish there were no validations of the attributes, because later we have a new attribute and if the library isnt updated the attribute wont be forwarded. I just tested this by adding a new made up attribute and its not written on the source. Only data-* attributes are written. I had this problem with react in the past had to comment the attribute validation because I needed an attribute for the iframe element that wasnt on react at the time. Is there a real need to validate attributes or can all of that be avoided? I guess this also happens with css properties?

**Comment by ryansolid** [comment]
I suspect I know what it is. It's setting the width property instead of the width attribute. It's possible they behave differently. I generally prefer properties over attributes for performance but I've made exceptions in the past where it makes sense. I have a lookup for specific attributes.

**Comment by ryansolid** [comment]
This might be a DOM/CSS-ism. I removed all the width properties and attributes and set the width attribute and although it reflected on the attribute as "100%" the element still had a width of 0. I created a new element and it was fine. Probably because I'd set the width property once. It looks like width is computed property. I'm gathering if I treat width and height as attributes we are probably in better shape here. I wonder if there is something I'm missing.

**Comment by titoBouzout** [comment]
You are right, the width and height are a setter and a getter, I missed that. Returns instance of Number which must be computed or something. setAttribute works as expected. 
If you want to make this really performant you may create one effect function for each attribute you set. Because you are triggering all the getters when any of the attributes/properties change which in some situations may cause multiple reflows(relevant to this property from a performance perspective).

**Comment by ryansolid** [comment]
Yeah so I should make these use setAttribute then I take it? I can make that change. I hadn't hit it before but I assume all HTMLElements have this behavior for width and height.

In terms of the effect thing, I'm not sure I follow. I do group these but I have guards, so it will only call the set on the attribute that has updated etc...

**Comment by titoBouzout** [comment]
> Yeah so I should make these use setAttribute then I take it?

 Yes, you will have to use `getAttribute `because `.width` in the example I have open now is giving me 110 so will never match '100%'. I think you can avoid this step by writing one effect function for each attribute instead.  I think is better to not read values that trigger layout changes, just write it for sure the browser must already check if the value is different or not.

>  I can make that change. I hadn't hit it before but I assume all HTMLElements have this behavior for width and height.

Im mostly sure all elements with a width attribute accept 100%. 

> In terms of the effect thing, I'm not sure I follow. I do group these but I have guards, so it will only call the set on the attribute that has updated etc...

The problem is that if you write width, the cache of the layout gets invalidated. If then you read width(just to check if the value changed) it  will cause a recalculation of the layout because the browser doesn't know the new dimensions. Once it knows it will cache again till you invalidate it again by writing. So the getters of the HTML Elements that change layout are /dangerous/ when you dont need them(because they also affect other elements).  This is not completely correct but will give you an idea:

```jsx
// bad (changing the dimensions of any element can cause a change on size to el1)
el1.width = 40 // invalidates cache
console.log(el1.width) // recalculates layout
el2.width = 40 // invalidates cache
console.log(el1.width) // recalculates layout
el3.width = 40 // invalidates cache
console.log(el1.width) // recalculates layout

// good
el1.width = 40 // invalidates cache
el2.width = 40 // invalidates cache
el3.width = 40 // invalidates cache

console.log(el1.width) // recalculates layout
console.log(el2.width) // layout is cached
```

There was a very old blog post explaining it nicely, cant find it maybe this helps 
https://gist.github.com/faressoft/36cdd64faae21ed22948b458e6bf04d5
and this one that also has a lot of related links
https://gist.github.com/paulirish/5d52fb081b3570c81e3a

**Comment by ryansolid** [comment]
Thanks for explaining. Yeah for sure I never read from the DOM. I'd hit this all over the place. I create an object that stores the previous objects and never read. Even though it is safe to do it at times all DOM interaction carries a cost so I avoid it like the plague. Leads to interesting seemingly more expensive workarounds but it is generally worth it. This shows what I mean: https://solid-template-explorer.netlify.app/#DwEwlgbgBAzg9gWwKYFoCGAXDAnAjAXgG8YNMkA6THXAX1kVSuwCYiSzKsW6B6APiA

(Technically it's Solid but it's really just showing DOM expressions compiled output so you might appreciate this as well).

And I will change width and height to attributes in the next release.

**Comment by ryansolid** [comment]
Ok I just released a new version of the babel plugin that should fix this. I need to do a MobX-JSX(all the libraries technically) release to fix the spread operator as well but that will come a bit later.

**Comment by titoBouzout** [comment]
aha! No wonder, very nice! Theres so much magic happening on modern js that is hard to keep track and is a bit unexpected sometimes. The example makes a lot of sense, and sure take your time. Thanks again

**Comment by ryansolid** [comment]
Should be all fixed in v0.12.3. Thanks for reporting.
